numpy/_core/src/umath/ufunc_type_resolution.c

```diff
@@ -2281,3 +2281,18 @@ PyUFunc_DivmodTypeResolver(PyUFuncObject *ufunc,
 
     return 0;
 }
+
+NPY_NO_EXPORT int
+PyUFunc_FindTypeResolver(PyUFuncObject *ufunc,
+                                NPY_CASTING casting,
+                                PyArrayObject **operands,
+                                PyObject *type_tup,
+                                PyArray_Descr **out_dtypes)
+{
+    out_dtypes[0] = NPY_DT_CALL_ensure_canonical(PyArray_DESCR(operands[0]));
+    out_dtypes[1] = NPY_DT_CALL_ensure_canonical(PyArray_DESCR(operands[1]));
+    out_dtypes[2] = NPY_DT_CALL_ensure_canonical(PyArray_DESCR(operands[2]));
+    out_dtypes[3] = NPY_DT_CALL_ensure_canonical(PyArray_DESCR(operands[3]));
+    out_dtypes[4] = PyArray_DescrFromType(NPY_LONG);
```

Comment:
Is this better? What should we be using if not `long`?

Reply:
You don't do any casts, you just make sure that the promotion step designates the loop you actual implement (for any reasonable input).  The promotion step is currently this legacy type resoltion (which is *not* used for the actual `resolve_dtypes`, you do not even need a `resolve_dtypes`).
NumPy defaults to `same-kind` casting, so it will just do the cast without a second thought.

Reply:
Actually, to not change the result dtype, it should be typed as long (and `intp` in the future).